x/upgrade/handler.go

```diff
@@ -0,0 +1,35 @@
+package upgrade
+
+import (
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	govtypes "github.com/cosmos/cosmos-sdk/x/gov/types"
+)
+
+// NewSoftwareUpgradeProposalHandler creates a governance handler to manage new proposal types.
+// It enables SoftwareUpgradeProposal to propose an Upgrade, and CancelSoftwareUpgradeProposal
+// to abort a previously voted upgrade.
+func NewSoftwareUpgradeProposalHandler(k Keeper) govtypes.Handler {
+	return func(ctx sdk.Context, content govtypes.Content) sdk.Error {
+		switch c := content.(type) {
+		case SoftwareUpgradeProposal:
+			return handleSoftwareUpgradeProposal(ctx, k, c)
+		case CancelSoftwareUpgradeProposal:
+			return handleCancelSoftwareUpgradeProposal(ctx, k, c)
```

Comment:
What happens when you cancel and a governance proposal is already mid-vote-period?

Reply:
Cancel is also a governance proposal. It will actually clear any existing plan (there can only be one).

I mean, I guess you can start a vote on a proposal on Friday, then Monday start a vote on Cancel (which will only take effect after the other vote it finished). Two weeks later on Friday, we have a plan set. The following Monday unset.

As long as the time between proposing upgrade and proposing cancel is less than the time between upgrade vote finishing and the planned upgrade block, you are fine. I guess this is a good reason to leave a week or two after the vote finishes for implementing it, so a following cancel upgrade proposal has a chance to be voted on.

If this makes sense, I can add it to docs (or godoc strings)